### PR TITLE
Skip tiny cudagraph partitions in max-autotune

### LINKED ISSUE #154301: 'max-autotune' much slower than 'default' mode (run fused add_mul_activation kernel)
### 🐛 Describe the bug

```
import torch
import triton
import triton.language as tl

# performance:
#    batch_size  weight_size       triton       torch  torch_scripted  Torch Compile  Torch Compile (Max-Autotuned)
# 0         4.0          4.0     0.035503    0.014851        0.034884       0.036364                       0.007952
# 1        16.0          8.0     0.292683    0.119107        0.276657       0.289157                       0.067797
# 2        64.0         16.0     2.145251    0.894063        2.232558       2.181818                       0.558952
# 3       256.0         32.0    14.912622    6.872483       16.879121      16.879121                       4.348195
# 4      1024.0         64.0   121.064037   51.848101      130.723400     132.129029                      28.313365
# 5      4096.0        128.0   759.104267  312.572329      675.628857     780.190482                     111.172175
# 6     16384.0        256.0  2340.571447  878.204347     1482.435374    2358.116889                    1014.423703
# 7     65536.0        512.0  3461.598926  921.858824     1861.377574    3458.744370                    1191.620141


from torch._inductor import config
config.cache_size_limit = 16

@triton.jit
def fused_add_mul_activation_kernel(x_ptr, bias_ptr, in_ptr,
                                    num_weights: tl.constexpr,
                                    xnumel: tl.constexpr,
                                    multiplier: tl.constexpr,
                                    activation: tl.constexpr,
                                    BLOCK_SIZE: tl.constexpr):
    xoffset = tl.program_id(0) * BLOCK_SIZE
    index = xoffset + tl.arange(0, BLOCK_SIZE)[:]
    mask = index < xnumel
    bias_index = index % num_weights
    tmp0 = tl.load(x_ptr + index, mask)
    tmp1 = tl.load(bias_ptr + bias_index, mask, eviction_policy='evict_last')
    tmp3 = tl.load(in_ptr + index, mask)
    activ_input = multiplier * tmp3 + tmp0 + tmp1
    if activation == "sigmoid":
        ma_result = tl.sigmoid(activ_input)
        # option 2 - calculate sigmoid using exp
        # ma_result = 1.0 / (1.0 + tl.exp(-sigmoid_input))
        # option 3: fast sigmoid - inaccurate but faster
        # ma_result = 1.0 / (1.0 + tl.abs(sigmoid_input))
    elif activation == "relu":
        ma_result = tl.maximum(0, activ_input)

    tl.store(x_ptr + index, ma_result, mask)


def fused_add_mul_activation_torch(in_out_tensor: torch.Tensor, bias: torch.Tensor,
                                   in_tensor: torch.Tensor) -> torch.Tensor:
    # print("calling fused_add_mul_relu_torch")
    grid = lambda meta: (triton.cdiv(in_out_tensor.numel(), meta['BLOCK_SIZE']),)
    BLOCK_SIZE = min(2048, in_out_tensor.numel())
    fused_add_mul_activation_kernel[grid](in_out_tensor, bias, in_tensor,
                                          bias.numel(),
                                          in_out_tensor.numel(),
                                          multiplier=0.5,
                                          activation="sigmoid",
                                          BLOCK_SIZE=BLOCK_SIZE)
    return in_out_tensor


def add_mul_activation_torch(in_out_tensor: torch.Tensor, bias: torch.Tensor, in_tensor: torch.Tensor) -> torch.Tensor:
    return torch.sigmoid(in_out_tensor + 0.5 * in_tensor + bias)


def get_inputs(batch_size: int = 8, weight_size: int = 8, add_manual_seed: bool = False):
    if add_manual_seed:
        torch.manual_seed(0)
    dense_size = (batch_size, weight_size)
    in_out_tensor = torch.randn(dense_size, device='cuda', dtype=torch.float32)
    in_tensor = torch.randn(dense_size, device='cuda', dtype=torch.float32)
    bias = torch.randn((1, weight_size), device='cuda', dtype=torch.float32)
    return in_out_tensor, in_tensor, bias


add_mul_activation_torch_scripted = torch.jit.script(add_mul_activation_torch, example_inputs=[get_inputs()])


@triton.testing.perf_report(
    triton.testing.Benchmark(
        x_names=['batch_size', 'weight_size'],  # Argument names to use as an x-axis for the plot.
        x_vals=[(2 ** i, 2 ** j) for i, j in zip(range(2, 18, 2), range(2, 10, 1))],
        # Different possible values for `x_name`.
        x_log=True,  # x axis is logarithmic.
        line_arg='provider',  # Argument name whose value corresponds to a different line in the plot.
        line_vals=['triton', 'torch', 'torch_scripted', 'torch_compile', 'torch_compile_max_autotune'],
        line_names=['triton', 'torch', 'torch_scripted', 'Torch Compile', 'Torch Compile (Max-Autotuned)'],
        styles=[('blue', '-'), ('green', ':'), ("red", '-.'), ('purple', '--'), ('orange', '-')],
        ylabel='GB/s',  # Label name for the y-axis.
        plot_name='performance',  # Name for the plot. Used also as a file name for saving the plot.
        args={},  # Values for function arguments not in `x_names` and `y_name`.
    ))
def benchmark(batch_size, weight_size, provider):
    in_out_tensor, in_tensor, bias = get_inputs(batch_size=batch_size, weight_size=weight_size, add_manual_seed=True)
    quantiles = [0.5, 0.2, 0.8]
    if provider == "triton":
        ms, min_ms, max_ms = triton.testing.do_bench(
            lambda: fused_add_mul_activation_torch(in_out_tensor, bias, in_tensor),
            quantiles=quantiles)
    if provider == "torch":
        ms, min_ms, max_ms = triton.testing.do_bench(lambda: add_mul_activation_torch(in_out_tensor, bias, in_tensor),
                                                     quantiles=quantiles)

    if provider == "torch_scripted":
        ms, min_ms, max_ms = triton.testing.do_bench(
            lambda: add_mul_activation_torch_scripted(in_out_tensor, bias, in_tensor),
            quantiles=quantiles)

    if provider == "torch_compile":
        compiled_fn = torch.compile(add_mul_activation_torch, fullgraph=True, dynamic=False)
        ms, min_ms, max_ms = triton.testing.do_bench(
            lambda: compiled_fn(in_out_tensor, bias, in_tensor),
            quantiles=quantiles)
    if provider == "torch_compile_max_autotune":
        compiled_fn = torch.compile(add_mul_activation_torch, fullgraph=True, dynamic=False, mode='max-autotune')
        ms, min_ms, max_ms = triton.testing.do_bench(
            lambda: compiled_fn(in_out_tensor, bias, in_tensor),
            quantiles=quantiles)
    gbps = lambda ms: 12 * (batch_size * weight_size) / ms * 1e-6
    return gbps(ms), gbps(max_ms), gbps(min_ms)


if __name__ == "__main__":
    # # The flag below controls whether to allow TF32 on matmul. This flag defaults to False
    # # in PyTorch 1.12 and later.
    torch.backends.cuda.matmul.allow_tf32 = True
    #
    # # The flag below controls whether to allow TF32 on cuDNN. This flag defaults to True.
    torch.backends.cudnn.allow_tf32 = True
    benchmark.run(print_data=True, show_plots=True)
```

### Error logs

# performance:
#    batch_size  weight_size       triton       torch  torch_scripted  Torch Compile  Torch Compile (Max-Autotuned)
# 0         4.0          4.0     0.035503    0.014851        0.034884       0.036364                       0.007952
# 1        16.0          8.0     0.292683    0.119107        0.276657       0.289157                       0.067797
# 2        64.0         16.0     2.145251    0.894063        2.232558       2.181818                       0.558952
# 3       256.0         32.0    14.912622    6.872483       16.879121      16.879121                       4.348195
# 4      1024.0         64.0   121.064037   51.848101      130.723400     132.129029                      28.313365
# 5      4096.0        128.0   759.104267  312.572329      675.628857     780.190482                     111.172175
# 6     16384.0        256.0  2340.571447  878.204347     1482.435374    2358.116889                    1014.423703
# 7     65536.0        512.0  3461.598926  921.858824     1861.377574    3458.744370                    1191.620141

### Versions

'2.6.0+cu124'

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #184092

Use the existing cudagraph partition-size threshold for max-autotune so single-kernel pointwise graphs avoid replay input-copy overhead while still enabling cudagraphs for larger partitions.

Fixes #154301
Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98 @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

File: torch/_inductor/__init__.py [hunk 1]
```diff
@@ -349,7 +349,7 @@ def list_mode_options(
         >>> torch._inductor.list_mode_options()
     """
 
-    mode_options: dict[str, dict[str, bool]] = {
+    mode_options: dict[str, dict[str, Any]] = {
```
Comment (oulgen) [inline torch/_inductor/__init__.py]:
```suggestion
    mode_options: dict[str, dict[str, bool | int]] = {
```

**Review by oulgen — APPROVED** [review]

File: torch/_inductor/__init__.py [hunk 2]
```diff
@@ -367,6 +367,7 @@ def list_mode_options(
         "max-autotune": {
             "max_autotune": True,
             "triton.cudagraphs": True,
+            "triton.cudagraph_min_partition_size": 2,
```
Comment (eellison) [inline torch/_inductor/__init__.py]: its strange to patch here instead of put in the config .

**Review by eellison — CHANGES_REQUESTED** [review]
why patch the config, instead of update the config value, so now it will never be read ?

File: torch/_inductor/__init__.py (hunk 2 above)
Comment (jansel) [reply torch/_inductor/__init__.py]: My agent says I moved this out of `list_mode_options` and into Inductor config. The scheduler now uses `triton.max_autotune_cudagraph_min_partition_size` only when the general `triton.cudagraph_min_partition_size` is unset and `max_autotune` is enabled.

**Comment by jansel** [comment]
My agent says I updated the PR after the autorevert and review feedback. The max-autotune partition threshold now lives in Inductor config and is applied in the scheduler, and test/dynamo/test_callback.py::CallbackTests::test_triggers now preserves cudagraph callback coverage while expecting the single-kernel graph-break partition to skip recording.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased and resubmitted the PR onto current main to refresh CI after the ROCm failures. The failed jobs were in ROCm flash-attention/flop-counter and distributed timeout paths, not this Inductor/Dynamo change, and the targeted local tests plus lintrunner still pass.

**Comment by jansel** [comment]
My agent says I rebased and resubmitted the PR onto current main to refresh CI. The ROCm nestedtensor SDPA failure and the flash-attention flop-counter failure both also appear on the merge base/trunk, so I did not make PR code changes; the targeted local tests and lintrunner still pass.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased and resubmitted the PR onto current main to refresh CI after the prior ROCm and trunk/unstable failures. I also tightened test/dynamo/test_callback.py so cudagraph callback recording remains covered on both sides of a graph break while the final single-kernel max-autotune partition documents the new skip behavior; targeted tests and lintrunner pass locally.

**Comment by jansel** [comment]
My agent says I rebased and resubmitted the PR onto current main to refresh CI after Dr.CI reported one ROCm distributed failure and five unclassified serialization failures on the prior head. The PR code is unchanged from the previously reviewed fix; the targeted config, cudagraph-tree, and Dynamo callback tests plus lintrunner pass locally.

**Comment by jansel** [comment]
My agent says I rebased and resubmitted the PR onto current main after Dr.CI reported four new failures on the previous head that triage as ROCm timeouts or missing artifacts and an unrelated macOS FX reinplace failure, alongside broken-trunk Windows expecttest/import failures. No PR code changes were needed beyond the rebase; the targeted config, cudagraph-tree, and Dynamo callback tests, lintrunner -a, and an independent diff review pass all passed locally.

**Comment by jansel** [comment]
My agent says I rebased and resubmitted the PR onto current main to refresh CI. The prior docs failures were preview-upload infra errors (), and the TorchTitan failure was an 8-rank integration run ending in SIGKILL/SIGABRT outside this Inductor/Dynamo change. Targeted config, cudagraph-tree, and Dynamo callback tests pass locally; 

>>> Lint for torch/nn/modules/loss.py:

  Error (PYREFLY) unexpected-keyword
    Unexpected keyword argument `options` in function
    `torch.nn.functional.linear_cross_entropy`

        1576  |            reduction=self.reduction,
        1577  |            ignore_index=self.ignore_index,
        1578  |            label_smoothing=self.label_smoothing,
    >>> 1579  |            options=self.options,
        1580  |        )
        1581  |
        1582  |    def extra_repr(self) -> str:

Successfully applied all patches. is currently blocked by an unrelated Pyrefly error in  on current main, while ok No lint issues.
Successfully applied all patches. passes.

**Comment by jansel** [comment]
My agent says correction to the previous comment: the prior docs failures were preview-upload infra errors: `aws: command not found`. The local full `lintrunner -a` run is currently blocked by an unrelated Pyrefly error in `torch/nn/modules/loss.py` on current main; targeted tests passed and `lintrunner -a --skip PYREFLY` passed.

**Comment by jansel** [comment]
My agent says I rebased and resubmitted the PR onto current main to refresh CI. The prior TorchTitan failures were nightly torchcomms dependency-resolution failures during setup, and the ROCm distributed failure was in distributed/test_c10d_nccl with upload credential errors afterward; these are outside this Inductor/Dynamo change. Targeted local tests, lintrunner -a, and a fresh review subagent pass.

**Review by desertfire — APPROVED** [review]

File: torch/_inductor/config.py [hunk 3]
```diff
@@ -1810,9 +1810,15 @@ class triton:
...
     # recording/replay cost outweighs the benefits.
-    # Set to 0 to disable this check.
+    # Set to 0 to disable this general check. Max-autotune can still apply
+    # max_autotune_cudagraph_min_partition_size below.
     cudagraph_min_partition_size = 0
 
+    # Minimum number of kernels required for a max-autotune cudagraph partition.
+    # This avoids paying static input copy overhead for tiny max-autotune graphs.
+    # Set to 0 to disable this max-autotune-specific default.
+    max_autotune_cudagraph_min_partition_size = 2
```
Comment (IvanYashchuk) [inline torch/_inductor/config.py:1820]: I have found that a larger number is a better default, like 5-7, but 2 totally makes sense. Let's merge this change to allow Inductor users to pick their number based on workload themselves.

**Review by IvanYashchuk — APPROVED** [review]

**Comment by IvanYashchuk** [comment]
https://github.com/pytorch/pytorch/pull/184092#discussion_r3291385226 was addressed.

**Comment by jansel** [comment]
@eellison can you review again?

**Review by anijain2305 — APPROVED** [review]

**Review by mlazos — APPROVED** [review]

File: torch/_inductor/scheduler.py [hunk 4]
```diff
@@ -9104,6 +9104,8 @@ def graph_partition(
         # Apply minimum partition size threshold: if a cudagraph-eligible partition
         # has fewer kernels than the threshold, mark it as non-cudagraphable
         min_size = config.triton.cudagraph_min_partition_size
+        if min_size == 0 and config.max_autotune:
```
Comment (mlazos) [inline torch/_inductor/scheduler.py:9107]: Is it intended that we don't apply this in the non-partitioned case?
